langchain/agents/agent_toolkits/openapi/planner.py

```diff
@@ -0,0 +1,375 @@
+        try:
+            data = json.loads(text)
+            response = self.requests_wrapper.post(data["url"], data["data"])
+            response = response[: self.response_length]
+            # print(f"Request tool input: {data}\nResponse: {response}")
+            return self.llm_chain.predict(
+                response=response, instructions=data["output_instructions"]
+            ).strip()
+        except Exception as e:
+            return repr(e)
+
+    async def _arun(self, text: str) -> str:
+        pass
+
+
+#
+# Orchestrator, planner, controller.
+#
+def _create_api_planner_tool(
+    api_spec: ReducedOpenAPISpec, llm: BaseLanguageModel = DEFAULT_LLM
+) -> Tool:
+    endpoint_descriptions = [
+        f"{name} {description}" for name, description, _ in api_spec.endpoints
+    ]
+    prompt = PromptTemplate(
+        template=API_PLANNER_PROMPT,
+        input_variables=["query"],
+        partial_variables={"endpoints": "- " + "- ".join(endpoint_descriptions)},
+    )
+    chain = LLMChain(llm=llm, prompt=prompt)
+    tool = Tool(
+        name=API_PLANNER_TOOL_NAME,
+        description=API_PLANNER_TOOL_DESCRIPTION,
+        func=chain.run,
+    )
+    return tool
+
+
+def _create_api_controller_agent(
+    api_url: str,
+    api_docs: str,
+    requests_wrapper: RequestsWrapper,
+    llm: BaseLanguageModel = DEFAULT_LLM,
+) -> AgentExecutor:
+    tools = [
+        RequestsGetToolWithSummary(
+            requests_wrapper=requests_wrapper, response_length=MAX_RESPONSE_LENGTH
+        ),
+        RequestsPostToolWithSummary(
+            requests_wrapper=requests_wrapper, response_length=MAX_RESPONSE_LENGTH
+        ),
+    ]
+    prompt = PromptTemplate(
+        template=API_CONTROLLER_PROMPT,
+        input_variables=["input", "agent_scratchpad"],
+        partial_variables={
+            "api_url": api_url,
+            "api_docs": api_docs,
+            "tool_names": ", ".join([tool.name for tool in tools]),
+            "tool_descriptions": "\n".join(
+                [f"{tool.name}: {tool.description}" for tool in tools]
+            ),
+        },
+    )
+    agent = ZeroShotAgent(
+        llm_chain=LLMChain(llm=llm, prompt=prompt),
+        allowed_tools=[tool.name for tool in tools],
+    )
+    return AgentExecutor.from_agent_and_tools(agent=agent, tools=tools, verbose=True)
+
+
+def _create_api_controller_tool(
+    api_spec: ReducedOpenAPISpec, requests_wrapper: RequestsWrapper
+) -> Tool:
+    """Expose controller as a tool. The tool is invoked with a plan from the planner, and dynamically
+    creates a controller agent with relevant documentation only to constrain the context.
+    """
+
+    base_url = api_spec.servers[0]["url"]  # TODO: do better.
+
+    def _create_and_run_api_controller_agent(plan_str: str) -> str:
+        pattern = r"\b(GET|POST|PUT|DELETE)\s+(/\S+)*"
+        matches = re.findall(pattern, plan_str)
+        endpoint_names = [f"{match[0]} {match[1]}" for match in matches]
```

Comment:
```suggestion
        endpoint_names = [f"{verb} {description}" for verb, description in matches]
```


Alternatively, could use named capture groups

```
In [21]:  pattern = r"\b(?P<verb>GET|POST|PUT|DELETE)\s+(?P<description>/\S+)*"
    ...: 

In [22]: next(re.finditer(pattern, "GET /album/{id}")).groupdict()
Out[22]: {'verb': 'GET', 'description': '/album/{id}'}
```